router2023/main.md

```diff
@@ -0,0 +1,454 @@
+
+Not so fast. BL1 (Primary Bootloader) is the most basic bootloader and all it can do is to load next stage bootloader BL2. The Boot ROM is hardwired or configured to know where BL2 is located in memory or storage.
+This location is typically specified by the SoC manufacturer or system designer. The Boot ROM's role is to initiate the boot process by loading the secondary bootloader stage, BL2, into memory and passing control to it.
+BL2 then takes over the boot process, sets up the secure environment, initializes necessary components, and proceeds to load and verify subsequent stages of the bootloader.
+
+What is BL2 and where to find it?
+
+I am not sure if this is a general rule for ARM based SoC but in our case BL2 is [ARM Trusted Firmware (ATF)](https://github.com/ARM-software/arm-trusted-firmware).
+ATF plays a crucial role in the secure boot sequence of ARM-based systems, establishing a secure and trusted foundation for the entire system.
+It initializes the security mechanisms, sets up the secure execution environment (Secure Monitor Mode), and verifies the integrity of subsequent bootloader stages.
+Finally, it loads the next stage bootloader - BL3.
+
+BL3 is further divided into:
+
+- BL31 (sometimes written as BL3-1)
+- optional BL32 that performs some additional verification (BL3-2)
+- normal world bootloader BL33 (BL3-3)
+
+BL33 is the final boot loader stage. It is usually a more feature-rich bootloader. BL3-3's responsibility is to initialize additional hardware, load the operating system kernel, and transfer control to the operating system for it to fully boot.
+A popular option for embedded systems is to use [U-Boot](https://u-boot.readthedocs.io/en/latest/index.html).
+
+_How does ATF know where to find BL3?_
+
+There are different strategies how that can be achieved afaik, but in our case we will compile U-Boot "into" ATF.
+
+Let's summarize what we've learned so far. There is BL1 burned into the SoC that will start ATF that will then call U-Boot which will finally load our operating system.
+
+Does it mean that we can start now?
+
+Not yet. Remember when I told you about the device-tree? U-Boot is a generic program that can be deployed on a wide variety of different devices. It achieves that by abstracting over the hardware with the help of device-tree definitions.
+That means that we need to provide our bpir3 specific device-tree definitions to U-Boot in order to compile it for our board. Apart from that we will also need a `def_config` file that configures
+how to build U-Boot. You can grab it from here: https://github.com/mtk-openwrt/u-boot/blob/mtksoc/configs/mt7986a_bpir3_sd_defconfig
+
+Here is a great tutorial that goes in details on how to do that "by hand": https://forum.banana-pi.org/t/tutorial-build-customize-and-use-mediatek-open-source-u-boot-and-atf/13785
+
+_Note that U-Boot and ATF need to be compiled for ARM, so you will either need to setup cross-compilation or have an ARM cpu._
```

Comment:
This would be a good point to mention that Nix makes cross-compilation _extremely_ easy. Most packages can be cross-compiled by simply prepending `pkgsCross.<target-arch>.` to the attribute to be built, e.g. you can cross-compile `vim` by building `pkgsCross.aarch64-multiplatform.vim`.

Reply:
As noted in another comment, yeah I should mention this somewhere. Either here or there. I will have to play with it on my own before.

Reply:
@Majiir I started looking into this and I found this repository https://github.com/nabam/nixos-rockchip which is a great evidence that it is certainly possible.
However, I was unable to replicate it in my case. I think that I know why, and fixing it would require some non-small effort to refactor how the current flake is written(basically changing a lot of stuff to nixos modules and getting rid of hardcoded system). 

I think it would be great to do it but I don't have time atm for this :( 

It is still possible that I am missing something obvious and in reality that is much easier than I anticipate. 

Since (as far as I understand), you have done it in the past, could you take a look and confirm or deny my thesis? 

Here is my attempt: https://github.com/ghostbuster91/nixos-router/pull/35

Reply:
On my system, it's as simple as:

```nix
{
  nixosConfigurations.myMachine = nixpkgs.lib.nixosSystem {
    system = "x86_64-linux";
    modules = [
      {
        nixpkgs.crossSystem.system = "armv7l-linux";
      }
      ./configuration.nix
    ];
  }
}
```

Now, I'm pretty sure that what I'm doing is **not** the right or most flexible way to configure this. I've heard it's better to use options like `buildSystem` and `hostSystem`. But it does work.